File: Assignment 5/Java/SplitFileClient.java
import java.io.*;
import java.net.Socket;
import java.nio.file.*;

public class SplitFileClient {
    private static final String SERVER_ADDRESS_1 = "localhost";
    private static final int SERVER_PORT_1 = 44094;
    private static final String SERVER_ADDRESS_2 = "localhost";
    private static final int SERVER_PORT_2 = 54094;

    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Usage: java SplitFileClient <put|get> <filename>");
            return;
        }

        String cmd = args[0];
        String filename = args[1];

        try {
            if ("put".equalsIgnoreCase(cmd)) {
                String filePart1 = splitFile(filename, 1);
                String filePart2 = splitFile(filename, 2);

                sendPart(filename, filePart1, SERVER_ADDRESS_1, SERVER_PORT_1);
                Files.delete(Paths.get(filePart1));

                sendPart(filename, filePart2, SERVER_ADDRESS_2, SERVER_PORT_2);
                Files.delete(Paths.get(filePart2));

                System.out.println("File successfully split and sent to servers.");
            } else if ("get".equalsIgnoreCase(cmd)) {
                String filePart1 = getPart(filename, SERVER_ADDRESS_1, SERVER_PORT_1, 1);
                String filePart2 = getPart(filename, SERVER_ADDRESS_2, SERVER_PORT_2, 2);

                String outputFilename = getMergedFilename(filename);
                mergeFiles(filePart1, filePart2, outputFilename);

                Files.delete(Paths.get(filePart1));
                Files.delete(Paths.get(filePart2));

                System.out.println("File successfully retrieved and merged: " + outputFilename);
            } else {
                System.out.println("Usage: java SplitFileClient <put|get> <filename>");
            }
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void sendPart(String filename, String partFilename, String serverAddress, int serverPort) throws IOException {
        try (Socket socket = new Socket(serverAddress, serverPort);
             OutputStream socketOutput = socket.getOutputStream();
             InputStream socketInput = socket.getInputStream();
             FileInputStream partFile = new FileInputStream(partFilename)) {

            socketOutput.write(("PUT:" + filename).getBytes());
            socketOutput.flush();

            BufferedReader reader = new BufferedReader(new InputStreamReader(socketInput));
            String response = reader.readLine();
            if (!"READY".equals(response.trim())) {
                throw new IOException("Server not ready for file content");
            }

            byte[] buffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = partFile.read(buffer)) != -1) {
                socketOutput.write(buffer, 0, bytesRead);
            }
            socketOutput.flush();
        }
    }

    private static String getPart(String filename, String serverAddress, int serverPort, int partNum) throws IOException {
        String partFilename = filename + "-part" + partNum + ".tmp";
        try (Socket socket = new Socket(serverAddress, serverPort);
             OutputStream socketOutput = socket.getOutputStream();
             InputStream socketInput = socket.getInputStream();
             FileOutputStream partFile = new FileOutputStream(partFilename)) {

            socketOutput.write(("GET:" + filename).getBytes());
            socketOutput.flush();

            byte[] buffer = new byte[1025];
            int bytesRead;
            while ((bytesRead = socketInput.read(buffer)) != -1) {
                if (new String(buffer, 0, bytesRead).contains("EOF")) {
                    int eofIndex = new String(buffer, 0, bytesRead).indexOf("EOF");
                    if (eofIndex > 1) {
                        partFile.write(buffer, 1, eofIndex - 1);
                    }
                    break;
                } else {
                    partFile.write(buffer, 1, bytesRead - 1);
                }
            }
        }
        return partFilename;
    }

    private static String splitFile(String filename, int partNum) throws IOException {
        File inputFile = new File(filename);
        String partFilename = filename + "-part" + partNum + ".tmp";
        try (FileInputStream inputFileStream = new FileInputStream(inputFile);
             FileOutputStream partFileStream = new FileOutputStream(partFilename)) {

            byte[] buffer = new byte[1];
            boolean writeToPart1 = (partNum == 1);

            while (inputFileStream.read(buffer) != -1) {
                if (writeToPart1) {
                    partFileStream.write(buffer);
                }
                writeToPart1 = !writeToPart1;
            }
        }
        return partFilename;
    }

    private static void mergeFiles(String part1, String part2, String outputFile) throws IOException {
        try (FileOutputStream outFile = new FileOutputStream(outputFile);
             FileInputStream partFile1 = new FileInputStream(part1);
             FileInputStream partFile2 = new FileInputStream(part2)) {

            byte[] buffer1 = new byte[1];
            byte[] buffer2 = new byte[1];
            int bytesRead1, bytesRead2;

            while (true) {
                bytesRead1 = partFile1.read(buffer1);
                bytesRead2 = partFile2.read(buffer2);

                if (bytesRead1 == -1 && bytesRead2 == -1) {
                    break;
                }

                if (bytesRead1 != -1) {
                    outFile.write(buffer1, 0, bytesRead1);
                }
                if (bytesRead2 != -1) {
                    outFile.write(buffer2, 0, bytesRead2);
                }
            }
        }
    }

    private static String getMergedFilename(String filename) {
        int extIndex = filename.lastIndexOf('.');
        if (extIndex == -1) {
            return filename + "-merged";
        }
        return filename.substring(0, extIndex) + "-merged" + filename.substring(extIndex);
    }
}

File: Assignment 5/Go/SplitFileClient.go
/**
 * SplitFileClient.go
 * ID : 20194094
 * Name : Yongmin Yoo
 **/

package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strings"
)

const SERVER_ADDRESS_1 = "localhost:44094"
const SERVER_ADDRESS_2 = "localhost:54094"

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run client.go <put|get> <filename>")
		return
	}

	cmd := os.Args[1]
	filename := os.Args[2]

	if cmd == "put" {
		filePart1, filePart2, err := splitFile(filename)
		if err != nil {
			os.Remove(filename + "-part1.tmp")
			os.Remove(filename + "-part2.tmp")
			exitError(fmt.Sprintf("Failed to split file - %s", err.Error()))
		}

		err = sendPart(filename, filePart1, SERVER_ADDRESS_1)
		os.Remove(filePart1)
		if err != nil {
			exitError(fmt.Sprintf("Failed to send Part 1 - %s", err.Error()))
		}

		err = sendPart(filename, filePart2, SERVER_ADDRESS_2)
		os.Remove(filePart2)
		if err != nil {
			exitError(fmt.Sprintf("Failed to send Part 2 - %s", err.Error()))
		}

		fmt.Println("File successfully split and sent to servers.")
	} else if cmd == "get" {
		filePart1, err := getPart(filename, SERVER_ADDRESS_1, 1)
		if err != nil {
			os.Remove(filename + "-part1.tmp")
			exitError(fmt.Sprintf("Failed to get Part 1 - %s", err.Error()))
		}

		filePart2, err := getPart(filename, SERVER_ADDRESS_2, 2)
		if err != nil {
			os.Remove(filename + "-part1.tmp")
			os.Remove(filename + "-part2.tmp")
			exitError(fmt.Sprintf("Failed to get Part 2 - %s", err.Error()))
		}

		outputFilename := getMergedFilename(filename)
		err = mergeFiles(filePart1, filePart2, outputFilename)
		os.Remove(filePart1)
		os.Remove(filePart2)
		if err != nil {
			exitError(fmt.Sprintf("Failed to merge files - %s", err.Error()))
		}

		fmt.Println("File successfully retrieved and merged:", outputFilename)
	} else {
		fmt.Println("Usage: go run SplitFileClient.go <put|get> <filename>")
	}
}

func sendPart(filename, partFilename, serverAddress string) error {
	serverConn, err := net.Dial("tcp4", serverAddress)
	if err != nil {
		return err
	}
	defer serverConn.Close()

	partFile, err := os.Open(partFilename)
	if err != nil {
		return err
	}
	defer partFile.Close()

	serverConn.Write([]byte(fmt.Sprintf("PUT:%s", filename)))
	response, err := bufio.NewReader(serverConn).ReadString('\n')
	if err != nil {
		return err
	}
	if strings.TrimSpace(response) != "READY" {
		return fmt.Errorf("Server not ready for file content")
	}

	fileBuffer := make([]byte, 1024)
	for {
		fileLength, err := partFile.Read(fileBuffer)
		if err != nil && err != io.EOF {
			return err
		}
		if fileLength == 0 {
			break
		}
		serverConn.Write(fileBuffer[:fileLength])
	}
	return nil
}

func getPart(filename, serverAddress string, partNum int) (string, error) {
	serverConn, err := net.Dial("tcp4", serverAddress)
	if err != nil {
		return "", err
	}
	defer serverConn.Close()

	partFilename := filename + fmt.Sprintf("-part%d.tmp", partNum)
	serverConn.Write([]byte(fmt.Sprintf("GET:%s", filename)))

	partFile, err := os.Create(partFilename)
	if err != nil {
		return "", err
	}
	defer partFile.Close()

	partFileBuffer := make([]byte, 1025)
	for {
		partFileLength, err := serverConn.Read(partFileBuffer)
		if err != nil && err != io.EOF {
			return "", err
		}
		if partFileLength == 0 {
			break
		}

		if strings.Contains(string(partFileBuffer[:partFileLength]), "EOF") {
			eofIndex := strings.Index(string(partFileBuffer[:partFileLength]), "EOF")
			if eofIndex > 0 {
				partFile.Write(partFileBuffer[1:eofIndex])
			}
			break
		} else if string(partFileBuffer[:partFileLength][:6]) == "NOFILE" {
			return "", errors.New("Server has an error with file")
		} else if string(partFileBuffer[:partFileLength][:5]) == "ERROR" {
			return "", errors.New("Server Returned an Error")
		} else {
			partFile.Write(partFileBuffer[1:partFileLength])
		}
	}

	return partFilename, nil
}

func splitFile(filename string) (string, string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return "", "", err
	}
	defer file.Close()

	partFile1, err := os.Create(filename + "-part1.tmp")
	if err != nil {
		return "", "", err
	}
	defer partFile1.Close()

	partFile2, err := os.Create(filename + "-part2.tmp")
	if err != nil {
		return "", "", err
	}
	defer partFile2.Close()

	fileBuffer := make([]byte, 1)
	isPart1 := true
	for {
		fileLength, err := file.Read(fileBuffer)
		if err != nil && err != io.EOF {
			return "", "", err
		}
		if fileLength == 0 {
			break
		}

		if isPart1 {
			partFile1.Write(fileBuffer[:fileLength])
		} else {
			partFile2.Write(fileBuffer[:fileLength])
		}
		isPart1 = !isPart1
	}

	return filename + "-part1.tmp", filename + "-part2.tmp", nil
}

func mergeFiles(part1, part2, outputFile string) error {
	outFile, err := os.Create(outputFile)
	if err != nil {
		return err
	}
	defer outFile.Close()

	partFile1, err := os.Open(part1)
	if err != nil {
		return err
	}
	defer partFile1.Close()

	partFile2, err := os.Open(part2)
	if err != nil {
		return err
	}
	defer partFile2.Close()

	partFile1Buffer := make([]byte, 1)
	partFile2Buffer := make([]byte, 1)
	for {
		partFile1Length, err1 := partFile1.Read(partFile1Buffer)
		partFile2Length, err2 := partFile2.Read(partFile2Buffer)

		if partFile1Length == 0 && partFile2Length == 0 {
			break
		}

		if partFile1Length > 0 {
			outFile.Write(partFile1Buffer[:partFile1Length])
		}
		if partFile2Length > 0 {
			outFile.Write(partFile2Buffer[:partFile2Length])
		}

		if err1 != nil && err1 != io.EOF {
			return err1
		}
		if err2 != nil && err2 != io.EOF {
			return err2
		}
	}

	return nil
}

func getMergedFilename(filename string) string {
	ext := filepath.Ext(filename)
	base := strings.TrimSuffix(filename, ext)
	return base + "-merged" + ext
}

func exitError(msg string) {
	_, err := fmt.Fprintf(os.Stderr, "Error: %s\n", msg)
	if err != nil {
		fmt.Printf("Error: %s\n", msg)
	}
	os.Exit(1)
}
